README.md

```diff
@@ -229,6 +243,253 @@ Benchee.run %{"Awesome stuff" => fn -> magic end }
 your_teardown()
 ```
 
+When might this be useful?
+* Seeding the database with data to be used by all benchmarking functions
+* Starting/shutting down a server, process, other dependencies
+
+#### What is a scenario?
+
+For the following discussions, it's important to know what benchee considers a "benchmarking scenario". It's the combination of one benchmarking function you specify and an input. So, given this benchmark:
+
+```elixr
+Benchee.run %{
+  "foo" => fn(input) -> ... end,
+  "bar" => fn(input) -> ... end
+}, inputs: %{
+  "input 1" => 1,
+  "input 2" => 2
+}
+```
+
+there are 4 scenarios: foo with input 1, foo with input 2, bar with input 1 and bar with input 2. A scenario entails the whole thing - warmup and actual run time.
+
+
+#### Hook configuration: global versus local
+
+So, how do we get going? Hooks can be defined either _globally_ as part of the configuration and will then be executed for every scenario or _locally_ for a specific benchmarking function. When defined _locally_ the scenarios will only be executed for the scenarios of that benchmarking function.
+
+**Global hooks**
+
+Global hooks are specified as part of the general benchee configuration:
+
+```elixr
+Benchee.run %{
+  "foo" => fn(input) -> ... end,
+  "bar" => fn(input) -> ... end
+}, inputs: %{
+  "input 1" => 1,
+  "input 2" => 2,
+  before_scenario: fn(input) -> ... end
+}
+```
+
+Here the `before_scenario` function will be executed for all 4 scenarios present in this benchmarking suite.
+
+**Local hooks**
+
+Local hooks are defined as part of the definition of the benchmarking function. Therefore in the initial map instead of just a single function a tuple can be provided where the bencharking function comes first and then a list of keyword options follows specifying the hooks to run:
+
+```elixr
+Benchee.run %{
+  "foo" => {fn(input) -> ... end, before_scenario: fn(input) -> ... end},
+  "bar" => fn(input) -> ... end
+}, inputs: %{
+  "input 1" => 1,
+  "input 2" => 2
+}
+```
+
+Here `before_scenario` is only run for the 2 scenarios associated with `"foo"`, i.e. foo with input 1 and foo with input 2. It is _not_ run for any `"bar"` benchmarks.
+
+#### before_scenario
+
+Is executed before every [scenario](#what-is-a-scenario) that it applies to considering global/local configuration. `before_scenario` takes the input of the scenario as an argument.
+
+If you haven't defined `inputs`, it is still called with an argument albeit the special input which is returned by `Benchee.Benchmark.no_input()`. This also holds true for `before_each` and `after_scenario`.
+
+The defined function **must return whatever it wants following hooks and the benchmarking function to consider the input**. This means, that there usually are 3 kinds of before_scenarios:
+
```

Comment:
Maybe pull the rules about inputs and return values out into their own section, since they apply generally to all hooks?

Reply:
It gets complicated to but them in their own section, or at least I haven't figured out well how to. If you got a good idea how to do it, please tell me :)

These rules only apply to `before_*` hooks... it's hard to group them together imo:

* *_scenario/*_each - similar point of execution, but totally different semantics
* before_*/after_* - similar semantics (more so for before_* than for after_*), but different points of execution (although they happen after each other)

That's why I settled on giving each their own section, might be easier to find stuff but not sure.